src/BoltOn.Data.CosmosDb/Extensions.cs

```diff
@@ -19,5 +21,15 @@ public static BoltOnOptions BoltOnCosmosDbModule(this BoltOnOptions boltOnOption
             boltOnOptions.BoltOnAssemblies(Assembly.GetExecutingAssembly());
             return boltOnOptions;
         }
+
+        public static IServiceCollection AddCosmosDbContext<TCosmosDbContext>(this IServiceCollection serviceCollection, CosmosDbConfiguration configuration)
+        where TCosmosDbContext : BaseCosmosDbContext
+        {
+            var instance = Activator.CreateInstance<TCosmosDbContext>();
```

Comment:
try not to use Activator.CreateInstance anywhere, it should be your last resort. how about something like this:

`services.AddSingleton<TCosmosDbContext>((ctx) =>
{
    return new TCosmosDbContext(cosmosDbOptions);
});`

just an idea, not sure whether this work with the generic type. checkout the implementation factories here https://joonasw.net/view/aspnet-core-di-deep-dive

Reply:
try this:

` serviceCollection.AddSingleton(typeof(TCosmosDbContext), svc =>
            {
                var cosmosDbContext = new TCosmosDbContext();
                cosmosDbContext.SetConfiguration(configuration);
                return cosmosDbContext;
            });`

how about renaming the SetConfiguration method to Init() or Configure()? and within the method, check whether the configuration is already set, and only if not set, initialize it... so that nobody could change the configuration

Reply:
@gokulm : found a easeir way when the generic type is declared with where new(), we can instantiate it then and there without using any thing.